cpp/src/parquet/column_writer.cc

```diff
@@ -405,6 +412,9 @@ class SerializedPageWriter : public PageWriter {
   std::shared_ptr<Encryptor> data_encryptor_;
 
   std::shared_ptr<ResizableBuffer> encryption_buffer_;
+
+  int32_t num_dict_pages_;
+  std::map<Encoding::type, int32_t> num_data_pages_;
```

Comment:
Hi @wesm, 
I agree with you that unordered_map is the most suitable data structure. In fact I initially used an unordered_map. Surprisingly I encountered following build failures:
 
1. R studio build failing (Centos7) (gcc 4.8.5 c++ compiler)
2. C++ build failure (MacOS 10.15) (clang 7.1.0 c++ compiler)
The error associated with using **Encoding::type** as a key in unordered_map (std::unordered_map<Encoding::type, int32_t> num_data_pages_).  The reason to the for build failure was c++ compiler unable to find hash for "Encoding::type". ( But this is not an issue in later compiler versions. In case of gcc: gcc>6)

Therefore I thought, using a std::map will simplify the solution rather than using an unordered_map with hash_function. (May be need to define operator == to compare keys in case of hash collision).